### ISSUE TITLE
Proposal: treat readonly static members of class as namespace members

### ISSUE THREAD

**Issue by shrinktofit** [microsoft/TypeScript#38803]
## Search Terms

static member; namespace; type

Maybe related issues: https://github.com/microsoft/TypeScript/issues/38207

## Suggestion
Readonly static members of class should act like namespace members. Especially, when that member is assigned as an alias of enumeration or other class.

## Use Cases

With the suggestion,
```ts
enum Color { red, }
class Image {
    public static readonly Color = Color; // Marked as readonly, so tsc can ensure it's an alias
}
```

The `Image.Color` should be an alias of enumeration `Color`. So we can use it when an enumeration may be used:
```ts
console.log(Image.Color.red); // OK
function f(color: Image.Color) { // OK: used as type #1
}
```
`#1` is invalid for now because tsc does not treat `Image.Color` as a type.
To use `Image.Color` as a type, we may supply a verbose declaration according to the declaration merging rule of TypeScript:
```ts
namespace Image {
    export type Color = (typeof Color)[keyof typeof Color]; // :(
}
```
It's verbose...

## Examples

Embed enumerations, classes into other classes.

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [ ] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by kraenhansen** [comment]
@RyanCavanaugh you've marked this as "Awaiting More Feedback". I'm curious what sort of feedback you're looking for?

Personally I've encountered this a couple of times and was now about to create an issue for it .. then I found this.

There's a couple of reasons I think it would be valuable to provide this as an alternative for the only know workaround:
- The namespace feature has historically been subject of scrutiny.
- The recommended config from typescript-eslint has [no-namespace](https://typescript-eslint.io/rules/no-namespace) set to "error" by default.
- Exporting types via namespaces require a type alias or weird syntax (see last code-block of the description above) to avoid [a circular reference](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEkoGdnwGIHtPwN4F8AoUJOeAOygFsRkAHKMBAIVj0Pk-gBcBPOhFhwBeDNgDchfEA).

All in all, I think this is an example of something that would provide great value at what I would expect to be little cost.

**Comment by RyanCavanaugh** [comment]
FWIW it would not be little cost; quite the opposite because this would change how binding occurs at a fundamental level (since classes currently *don't* create a namespace meaning).

Regarding feedback, generally we're looking for both volume and severity, meaning that ideally lots of people are effectively blocked by some lack of feature. 1 heart and 1 comment in 2 years, when it's clearly a "nice to have" = this is extremely unlikely to happen.

**Comment by kraenhansen** [comment]
I think more and more are moving to named ESM exports over namespaces, so I can see why this issues hasn't been getting too much attention.

For more context, we're reimplementing an a package that was heavily relying on namespaces and this is the sort of code we see ourselves writing in the process: https://github.com/realm/realm-js/blob/c8c1a52d7a5a1bffd7f50e253f0d7367ab5d8be6/packages/realm/src/Realm.ts#L704-L734 (note the unfortunate duplication of type names and template arguments).
It would be much simpler to just declare these as static (read-only) members on the class: https://github.com/realm/realm-js/blob/c8c1a52d7a5a1bffd7f50e253f0d7367ab5d8be6/packages/realm/src/Realm.ts#L74-L85
